matrix/app_server/vision/optical_flow.py

```diff
@@ -0,0 +1,213 @@
+                # 2. Convert to NumPy array
+                # 3. OpenCV expects BGR, but since we convert to grayscale, RGB vs BGR doesn't matter.
+                frame_numpy = frame_tensor.permute(1, 2, 0).numpy().astype(np.uint8)
+                gray = cv2.cvtColor(frame_numpy, cv2.COLOR_RGB2GRAY)
+
+                if prev_gray is not None:
+                    flow = cv2.calcOpticalFlowFarneback(
+                        prev_gray, gray, None, 0.5, 3, 15, 3, 5, 1.2, 0
+                    )
+
+                    if self.return_flow:
+                        optical_flows.append(flow)
+
+                    magnitude, _ = cv2.cartToPolar(flow[..., 0], flow[..., 1])
+
+                    if self.calc_motion_score:
+                        motion_scores.append(float(np.mean(magnitude)))
+
+                prev_gray = gray
+
+        scores: Dict[str, float | List[List[float]]] = {}
+
+        if self.calc_motion_score:
+            scores["motion_score"] = (
+                float(np.mean(motion_scores)) if motion_scores else 0.0
+            )
+
+        if self.return_flow:
+            scores["flow"] = [flow.tolist() for flow in optical_flows]
+
+        return scores
+
+    @app.post("/run")
+    async def run(self, request: Request) -> JSONResponse:
+
+        if request.method == "POST":
+            request_json = await request.json()
+            timeout = request_json.get("timeout", TIMEOUT)
+
+        else:
+            return JSONResponse(
+                {"error": "Data must be in JSON format!"}, status_code=400
+            )
+
+        try:
+            if "video" not in request_json:
+                raise ValueError("Request must contain 'video' field with 'file_path'.")
+
+            video_path = request_json["video"].get("file_path")
+            if video_path is None:
+                raise ValueError("Request must contain 'file_path' field in 'video'.")
+
+            start: int = request_json["video"].get("start", 0)
+            end: Optional[int] = request_json["video"].get("end", None)
+            mode: SamplingMode = request_json["video"].get("mode", SamplingMode.FRAMES)
+            sampling_rate: Optional[float] = request_json["video"].get(
+                "sampling_rate", None
+            )
+            stride: Union[int, float] = request_json["video"].get("stride", 1)
+
+            dataset = TorchCodecVideoDataset(
+                video_path=video_path,
+                start=start,
+                end=end,
+                mode=mode,
+                stride=stride,
+                sampling_rate=sampling_rate,
+            )
+            data_loader = DataLoader(
+                dataset=dataset,
+                batch_size=self.batch_size,
+                shuffle=False,
+                num_workers=self.num_workers,
+                persistent_workers=self.num_workers > 0,
+            )
+
+            result = self._calculate_motion_metrics(
+                data_loader,
+            )
```

Comment:
use asyncio.to_threads() to avoid blocking

Reply:
The dataloader takes care of efficient loading and the task itself is mostly CPU-bound, so I think threading might not be helpful here, wdyt?